File: artifacts/api-server/src/routes/jobs.ts
import { Router, type IRouter } from "express";
import { db, jobsTable } from "@workspace/db";
import { eq, count, desc } from "drizzle-orm";
import { z } from "zod";

const router: IRouter = Router();

function formatJob(j: typeof jobsTable.$inferSelect) {
  return {
    ...j,
    startedAt: j.startedAt?.toISOString() ?? null,
    completedAt: j.completedAt?.toISOString() ?? null,
    createdAt: j.createdAt.toISOString(),
  };
}

const createJobBody = z.object({
  type: z.enum(["fetch", "process", "upload", "sync"]),
  storeId: z.number().nullable().optional(),
  config: z.record(z.unknown()).nullable().optional(),
});

router.get("/", async (req, res) => {
  try {
    const page = Math.max(1, parseInt(String(req.query.page || "1")));
    const limit = Math.min(100, Math.max(1, parseInt(String(req.query.limit || "20"))));
    const offset = (page - 1) * limit;

    const [jobs, [{ total }]] = await Promise.all([
      db.select().from(jobsTable).orderBy(desc(jobsTable.createdAt)).limit(limit).offset(offset),
      db.select({ total: count() }).from(jobsTable),
    ]);

    res.json({
      jobs: jobs.map(formatJob),
      total: Number(total),
      page,
      limit,
      totalPages: Math.ceil(Number(total) / limit),
    });
  } catch (err) {
    req.log.error({ err }, "Failed to list jobs");
    res.status(500).json({ error: "Internal server error" });
  }
});

router.post("/", async (req, res) => {
  try {
    const body = createJobBody.parse(req.body);

    const [job] = await db.insert(jobsTable).values({
      type: body.type,
      status: "pending",
      storeId: body.storeId ?? null,
      config: body.config ?? null,
      totalItems: 0,
      processedItems: 0,
      failedItems: 0,
      progressPercent: 0,
    }).returning();

    // Simulate starting the job asynchronously
    void simulateJobExecution(job!.id, body.type);

    res.status(201).json(formatJob(job!));
  } catch (err) {
    req.log.error({ err }, "Failed to create job");
    res.status(400).json({ error: "Invalid input" });
  }
});

router.get("/:id", async (req, res) => {
  try {
    const id = parseInt(req.params.id!);
    const [job] = await db.select().from(jobsTable).where(eq(jobsTable.id, id));
    if (!job) return res.status(404).json({ error: "Job not found" });
    res.json(formatJob(job));
  } catch (err) {
    req.log.error({ err }, "Failed to get job");
    res.status(500).json({ error: "Internal server error" });
  }
});

router.post("/:id/cancel", async (req, res) => {
  try {
    const id = parseInt(req.params.id!);
    const [job] = await db.select().from(jobsTable).where(eq(jobsTable.id, id));
    if (!job) return res.status(404).json({ error: "Job not found" });

    if (!["pending", "running"].includes(job.status)) {
      return res.status(400).json({ error: "Job cannot be cancelled in its current state" });
    }

    const [updated] = await db.update(jobsTable)
      .set({ status: "cancelled", completedAt: new Date() })
      .where(eq(jobsTable.id, id))
      .returning();

    res.json(formatJob(updated!));
  } catch (err) {
    req.log.error({ err }, "Failed to cancel job");
    res.status(500).json({ error: "Internal server error" });
  }
});

async function simulateJobExecution(jobId: number, type: string) {
  await new Promise((r) => setTimeout(r, 500));

  try {
    await db.update(jobsTable).set({
      status: "running",
      startedAt: new Date(),
      totalItems: 50,
    }).where(eq(jobsTable.id, jobId));

    // Check if cancelled
    const checkAndUpdate = async (processed: number) => {
      const [job] = await db.select().from(jobsTable).where(eq(jobsTable.id, jobId));
      if (!job || job.status === "cancelled") return false;

      await db.update(jobsTable).set({
        processedItems: processed,
        progressPercent: (processed / 50) * 100,
      }).where(eq(jobsTable.id, jobId));
      return true;
    };

    for (let i = 10; i <= 50; i += 10) {
      await new Promise((r) => setTimeout(r, 800));
      const cont = await checkAndUpdate(i);
      if (!cont) return;
    }

    await db.update(jobsTable).set({
      status: "completed",
      completedAt: new Date(),
      processedItems: 50,
      progressPercent: 100,
    }).where(eq(jobsTable.id, jobId));
  } catch {
    await db.update(jobsTable).set({
      status: "failed",
      completedAt: new Date(),
      errorMessage: "Simulation error",
    }).where(eq(jobsTable.id, jobId));
  }
}

export default router;

File: artifacts/pipeline/routers/jobs.py
from datetime import datetime, timezone
import math
from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy import select, func
from database import get_db
from models.models import Job, JobStatus, JobType, JobLog
from schemas.schemas import JobCreate, JobOut, JobListOut

router = APIRouter(prefix="/jobs", tags=["jobs"])


@router.get("", response_model=JobListOut)
async def list_jobs(
    page: int = Query(1, ge=1),
    limit: int = Query(20, ge=1, le=100),
    type: str = Query(None, description="Filter by job type: fetch|process|upload|sync"),
    status: str = Query(None, description="Filter by status: pending|running|completed|failed"),
    db: AsyncSession = Depends(get_db),
):
    q = select(Job)
    if type:
        try:
            q = q.where(Job.type == JobType(type))
        except ValueError:
            pass
    if status:
        try:
            q = q.where(Job.status == JobStatus(status))
        except ValueError:
            pass

    count_q = select(func.count()).select_from(q.subquery())
    total = (await db.execute(count_q)).scalar_one()

    jobs = (
        await db.execute(
            q.order_by(Job.created_at.desc()).offset((page - 1) * limit).limit(limit)
        )
    ).scalars().all()

    return JobListOut(
        jobs=[JobOut.model_validate(j) for j in jobs],
        total=total,
        page=page,
        limit=limit,
        total_pages=max(1, math.ceil(total / limit)),
    )


@router.post("", response_model=JobOut)
async def create_job(body: JobCreate, db: AsyncSession = Depends(get_db)):
    try:
        job_type = JobType(body.type)
    except ValueError:
        raise HTTPException(400, f"Invalid job type: {body.type}")

    # Validate source job exists if provided
    if body.source_job_id:
        source = await db.get(Job, body.source_job_id)
        if not source:
            raise HTTPException(400, f"Source job #{body.source_job_id} not found")

    job = Job(
        type=job_type,
        status=JobStatus.pending,
        store_id=body.store_id,
        config=body.config or {},
        source_job_id=body.source_job_id,
        started_at=datetime.now(timezone.utc),
    )
    db.add(job)
    await db.commit()
    await db.refresh(job)

    # Dispatch to Celery worker
    from tasks.job_tasks import run_job
    run_job.delay(job.id)

    return JobOut.model_validate(job)


@router.get("/{job_id}", response_model=JobOut)
async def get_job(job_id: int, db: AsyncSession = Depends(get_db)):
    job = await db.get(Job, job_id)
    if not job:
        raise HTTPException(404, "Job not found")
    return JobOut.model_validate(job)


@router.get("/{job_id}/logs")
async def get_job_logs(
    job_id: int,
    limit: int = Query(100, ge=1, le=500),
    db: AsyncSession = Depends(get_db),
):
    job = await db.get(Job, job_id)
    if not job:
        raise HTTPException(404, "Job not found")

    logs = (
        await db.execute(
            select(JobLog)
            .where(JobLog.job_id == job_id)
            .order_by(JobLog.created_at.asc())
            .limit(limit)
        )
    ).scalars().all()

    return {
        "job_id": job_id,
        "logs": [
            {
                "id": log.id,
                "level": log.level,
                "message": log.message,
                "created_at": log.created_at.isoformat() if log.created_at else None,
            }
            for log in logs
        ],
    }


@router.post("/{job_id}/cancel", response_model=JobOut)
async def cancel_job(job_id: int, db: AsyncSession = Depends(get_db)):
    job = await db.get(Job, job_id)
    if not job:
        raise HTTPException(404, "Job not found")
    if job.status not in (JobStatus.pending, JobStatus.running):
        raise HTTPException(400, "Job cannot be cancelled")

    job.status = JobStatus.cancelled
    job.completed_at = datetime.now(timezone.utc)
    await db.commit()
    await db.refresh(job)
    return JobOut.model_validate(job)
